PHP 8 support

Ayesh opened the pull request:
- Allow PHP 8 in `composer.json`
 - Allow PHPUnit 7.0 through 9.3. PHPUnit 9.3 supports PHP 8.0 and is the only official lowest version to do so. 
 - Prophecy is currently not allowing PHP 8.0, hence this being a draft PR.

Ayesh committed:
Add support for PHP 8 and tests

l0gicgate commented on .travis.yml:
```diff
@@ -10,12 +10,14 @@ matrix:
 before_script:
   - if [[ "$ANALYSIS" == 'true' ]]; then composer require php-coveralls/php-coveralls:^2.1.0 ; fi
-  - composer install -n
+  - if [[ "IGNOREPHP" == 'true' ]]; then composer install -n --ignore-platform-reqs; fi
+  - if [[ "IGNOREPHP" != 'true' ]]; then composer install -n; fi
```
Why do we need to ignore platform requirements? Is there another way to fix this?

Ayesh commented:
Hi @l0gicgate - thanks a lot for reviewing this. 
This PR was made before PHPUnit 9 had PHP 8 support in composer.json, and before PHPUnit 8.5 fixed compatibility issues with PHP 8. So I had to use `ignore-platform-reqs` to bypass Composer validating them. 

I removed them, and will mark this PR as ready. 

Travis internally does not provide PHP 8 under `trusty` (14 LTS), so I also updated to `xenial` (16 LTS) for PHP 8 builds, which now supports all PHP versions we test against. 

All tests now pass as well. 

Thank you.

akrabat commented:
Thanks for your work on this. Due to our tardiness, there were multiple PRs raised and I picked #71.
